### PR TITLE
Telemetry consistent for app service in eclipse

### PR DESCRIPTION
**Pull request by zmssp**
This is a part of #2874, include the telemetry util and app service in eclipse. 
Here is one example for the new telemetry.

image

### CODE REVIEW

File: Utils/azuretools-core/src/com/microsoft/azuretools/telemetry/AppInsightsClient.java [hunk 1]
```diff
@@ -175,4 +212,62 @@ public static void createFTPEvent(String eventName, String uri, String appName,
...
+    public static void sendOpEnd(EventType eventType, String operName, Map<String, String> properties) {
+        sendOpEnd2(eventType, operName, properties, null);
+    }
+
+    public static void sendOpEnd(EventType eventType, String operName, Map<String, String> properties,
+        Map<String, Double> metrics) {
+        sendOpEnd2(eventType, operName, properties, metrics);
+    }
+
+    private static void sendOpEnd2(EventType eventType, String operName, Map<String, String> properties,
```
Comment (andxu) [inline Utils/azuretools-core/src/com/microsoft/azuretools/telemetry/AppInsightsClient.java]: bad name

File: Utils/azuretools-core/src/com/microsoft/azuretools/telemetry/AppInsightsClient.java [hunk 2]
```diff
@@ -175,4 +212,62 @@ public static void createFTPEvent(String eventName, String uri, String appName,
...
+        Map<String, String> properties) {
+        sendError2(eventType, operName, errorType, errMsg, properties, null);
+    }
+
+    public static void sendError(EventType eventType, String operName, ErrorType errorType, String errMsg,
+        Map<String, String> properties, Map<String, Double> metrics) {
+        sendError2(eventType, operName, errorType, errMsg, properties, metrics);
+    }
+
+    private static void sendError2(EventType eventType, String operName, ErrorType errorType, String errMsg,
```
Comment (andxu) [inline Utils/azuretools-core/src/com/microsoft/azuretools/telemetry/AppInsightsClient.java]: bad name

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.webapp/src/com/microsoft/azuretools/webapp/ui/AppServiceCreateDialog.java [hunk 3]
```diff
@@ -1093,6 +1099,8 @@ protected void okPressed() {
                         Display.getDefault().asyncExec(() -> AppServiceCreateDialog.super.cancelPressed());
                     }
                     try {
+                        TelemetryUtil.sendTelemetryOpStart("create web app", properties);
+                        long timeS = System.currentTimeMillis();
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.webapp/src/com/microsoft/azuretools/webapp/ui/AppServiceCreateDialog.java]: time, timeS is not readable.

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.webapp/src/com/microsoft/azuretools/webapp/ui/WebAppDeployDialog.java [hunk 4]
```diff
@@ -177,6 +179,7 @@ protected Control createDialogArea(Composite parent) {
             @Override
             public void widgetSelected(SelectionEvent e) {
                 sendTelemetry("CREATE");
+                TelemetryUtil.sendTelemetryOpEnd("open create webapp diag", buildProperties());
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.webapp/src/com/microsoft/azuretools/webapp/ui/WebAppDeployDialog.java]: why two lines of temeletry?

Comment (zmssp) [reply PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.webapp/src/com/microsoft/azuretools/webapp/ui/WebAppDeployDialog.java]: One is the new telemetry schema, another is the old one. Should keep the old one for one or two release version

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.webapp/src/com/microsoft/azuretools/webapp/ui/WebAppDeployDialog.java [hunk 5]
```diff
@@ -587,6 +594,9 @@ protected IStatus run(IProgressMonitor monitor) {
                 String successMessage = "";
                 String errorMessage = "Error";
                 Map<String, String> postEventProperties = new HashMap<>();
+                postEventProperties.put("runtime",
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.webapp/src/com/microsoft/azuretools/webapp/ui/WebAppDeployDialog.java]: why runtime and duration key is lower-cased?  are them special?

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.webapp/src/com/microsoft/azuretools/webapp/ui/AppServiceCreateDialog.java (hunk 3 above)
Comment (zmssp) [reply PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.webapp/src/com/microsoft/azuretools/webapp/ui/AppServiceCreateDialog.java]: timeS means timeStart, or just use timeStart

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.webapp/src/com/microsoft/azuretools/webapp/ui/WebAppDeployDialog.java (hunk 5 above)
Comment (zmssp) [reply PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.webapp/src/com/microsoft/azuretools/webapp/ui/WebAppDeployDialog.java]: The data in kusto is all lowercase, so when we collect data just give a lower case name

File: Utils/azuretools-core/src/com/microsoft/azuretools/telemetry/AppInsightsClient.java (hunk 2 above)
Comment (zmssp) [reply Utils/azuretools-core/src/com/microsoft/azuretools/telemetry/AppInsightsClient.java]: Hard to give a nice name, I will give it a new name

File: Utils/azuretools-core/src/com/microsoft/azuretools/telemetry/AppInsightsClient.java [hunk 6]
```diff
@@ -47,6 +54,17 @@
...
     }
 
+    public enum EventName {
+        opStart,
+        opEnd,
+        error
+    }
+
+    public enum ErrorType {
+        userError,
```
Comment (Eskibear) [inline Utils/azuretools-core/src/com/microsoft/azuretools/telemetry/AppInsightsClient.java]: inconsistency names

File: Utils/azuretools-core/src/com/microsoft/azuretools/telemetry/AppInsightsClient.java [hunk 7]
```diff
@@ -25,13 +25,20 @@
...
 import com.microsoft.azuretools.azurecommons.helpers.Nullable;
-
 import java.util.HashMap;
 import java.util.Iterator;
 import java.util.Map;
 import java.util.UUID;
 
 public class AppInsightsClient {
+    private static final String EVENT_SPLIT = "/";
+    private static final String OPERATION_ID = "operationid";
```
Comment (Eskibear) [inline Utils/azuretools-core/src/com/microsoft/azuretools/telemetry/AppInsightsClient.java]: keep consistency in prop names, i.e. keep using camel case.

Comment (Eskibear) [reply Utils/azuretools-core/src/com/microsoft/azuretools/telemetry/AppInsightsClient.java]: although the keys get transformed to the lower case, it's better to keep the original data consistency, and for better readability. (see your own screenshot)

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.webapp/src/com/microsoft/azuretools/webapp/ui/WebAppDeployDialog.java [hunk 8]
```diff
@@ -674,15 +687,18 @@ protected IStatus run(IProgressMonitor monitor) {
...
                     LOG.log(new Status(IStatus.ERROR, Activator.PLUGIN_ID,
                         "run@ProgressDialog@deploy@AppServiceCreateDialog", ex));
                     AzureDeploymentProgressNotification.notifyProgress(this, deploymentName, null, -1, errorMessage);
                     webApp.start();
                     Display.getDefault().asyncExec(() -> ErrorWindow.go(parentShell, ex.getMessage(), errTitle));
+                    TelemetryUtil.sendTelemetryOpError("Deploy webapp", ErrorType.SystemError, ex.getMessage(),
+                        postEventProperties);
                 }
-                AppInsightsClient.create("Deploy as WebApp", "", postEventProperties);
+
```
Comment (Eskibear) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.webapp/src/com/microsoft/azuretools/webapp/ui/WebAppDeployDialog.java]: remove the empty line

File: Utils/azuretools-core/src/com/microsoft/azuretools/telemetry/AppInsightsClient.java [hunk 9]
```diff
@@ -25,13 +25,20 @@
...
+    private static final String OPERATION_NAME = "operationname";
+    private static final String ERROR_CODE = "errorCode";
+    private static final String ERROR_MSG = "message";
+    private static final String ERROR_TYPE = "errorType";
+
```
Comment (Eskibear) [inline Utils/azuretools-core/src/com/microsoft/azuretools/telemetry/AppInsightsClient.java]: no need for consequent empty lines here

File: Utils/azuretools-core/src/com/microsoft/azuretools/telemetry/AppInsightsClient.java (hunk 7 above)
Comment (zmssp) [reply Utils/azuretools-core/src/com/microsoft/azuretools/telemetry/AppInsightsClient.java]: agree, done

**Comment by Eskibear** [comment]
One question is, when I'm write a piece of code for an operation, how am I able to know the operationId if I want to send some extra info telemetry?

File: PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.webapp/src/com/microsoft/azuretools/webapp/ui/WebAppDeployDialog.java [hunk 10]
```diff
@@ -189,6 +192,7 @@ public void widgetSelected(SelectionEvent e) {
             @Override
             public void widgetSelected(SelectionEvent e) {
                 sendTelemetry("DELETE");
+                TelemetryUtil.sendTelemetryOpEnd("delete app service", buildProperties());
```
Comment (Flanker32) [inline PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.webapp/src/com/microsoft/azuretools/webapp/ui/WebAppDeployDialog.java]: seems that "CREATE"->"open create webapp dialog" and "DELETE"->"delete app service", is there any traces for create app service

File: Utils/azuretools-core/src/com/microsoft/azuretools/telemetry/AppInsightsClient.java [hunk 11]
```diff
@@ -106,38 +122,57 @@ public static void create(String eventName, String version, @Nullable Map<String
...
+            Map.Entry<String, String> entry = iter.next();
+            if (StringUtils.isNullOrEmpty(entry.getKey()) || StringUtils.isNullOrEmpty(entry.getValue())) {
+                iter.remove();
+            }
+        }
+        if (version != null && !version.isEmpty()) {
+            properties.put("Library Version", version);
+        }
+        String pluginVersion = configuration.pluginVersion();
+        if (pluginVersion != null && !pluginVersion.isEmpty()) {
```
Comment (Flanker32) [inline Utils/azuretools-core/src/com/microsoft/azuretools/telemetry/AppInsightsClient.java]: use `StringUtils.isNullOrEmpty` here?

Comment (zmssp) [reply Utils/azuretools-core/src/com/microsoft/azuretools/telemetry/AppInsightsClient.java]: good suggestion, done

**Comment by Flanker32** [comment]
LGTM

**Pull request merged**